GitUI/CommandsDialogs/FormCommit.cs

```diff
@@ -1325,6 +1332,16 @@ void DoCommit()
                     return;
                 }
 
+                GPGKeysUIController gpg = toolStripGpgKeyComboBox.KeysUIController;
+                if (!gpg.ValidateCommitSign(gpgSignCommitToolStripComboBox.SelectedIndex > 0, toolStripGpgKeyComboBox.KeyID))
+                {
+                    MessageBox.Show(this, TranslatedStrings.InvalidGpgSignOptions, _invalidSignCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Asterisk);
+                    toolStripMenuItem3.ShowDropDown();
+                    gpgSignCommitToolStripComboBox.Focus();
+
+                    return;
+                }
```

Comment:
This looks as though we're leaking implementation details. This should be a validation event on the combobox itself.

Reply:
It is a validation of 2 fields.  Index 0 is Do not sign Commit 1 and 2 are sign with default or sign with selected key.  This validation is a you can't commit with these 2 values of these fields.  The cases this covers are

- User has configured the commit.gpgSign value to true but does not have a default key so the key combobox will be no key selected.  Since git config is not fully in our hands this state is possible
    - would result in state of `git commit ... -S` along with the user.signingKey not being set. Git would complain and fail the commit.  
- User has selected sign with specific key and then failed to select a key.
![image](https://github.com/gitextensions/gitextensions/assets/3629489/1ac1dd91-02d2-47a7-b504-32d592d1de92)

![invalid selection](https://github.com/gitextensions/gitextensions/assets/3629489/a51ac030-606a-4e6f-800a-cf33dac4f044)
User can skip the options menu completely.  So I disagree with the validation event.  Where the validation needs to occur is  when they hit commit.  
The validation logic is the same for tag.gpgSign also.  Along with the validation being coded once in the controller and the ui needing to use it can do what it needs when it detects the state.